plutus-ledger-api/src/v1/value.rs

```diff
@@ -30,22 +55,64 @@ pub enum CurrencySymbol {
     NativeToken(MintingPolicyHash),
 }
 
+impl CurrencySymbol {
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, ConversionError> {
+        if bytes.is_empty() {
+            Ok(CurrencySymbol::Ada)
+        } else {
+            Ok(CurrencySymbol::NativeToken(MintingPolicyHash::from_bytes(
+                bytes,
+            )?))
+        }
+    }
+}
+
+impl fmt::Display for CurrencySymbol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            CurrencySymbol::Ada => {
+                if f.alternate() {
+                    write!(f, "lovelace")
+                } else {
+                    write!(f, "")
+                }
+            }
+            CurrencySymbol::NativeToken(symbol) => write!(f, "{}", symbol.0 .0),
+        }
+    }
+}
+
+impl FromStr for CurrencySymbol {
+    type Err = ConversionError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        all_consuming(currency_symbol)(s)
+            .finish()
+            .map_err(|err| {
+                ConversionError::ParseError(anyhow!(
+                    "Error while parsing CurrencySymbol '{}': {}",
+                    s,
+                    err
+                ))
+            })
+            .map(|(_, cs)| cs)
+    }
+}
+
 impl IsPlutusData for CurrencySymbol {
     fn to_plutus_data(&self) -> PlutusData {
         match self {
-            CurrencySymbol::Ada => String::from("").to_plutus_data(),
             CurrencySymbol::NativeToken(policy_hash) => policy_hash.to_plutus_data(),
+            CurrencySymbol::Ada => String::from("").to_plutus_data(),
```

Comment:
I kinda hate this. We are relying on the internal implementation of `IsPlutusData` instance of `String` while `CurrencySymbol` is in fact not a `String`.

Reply:
I just gave it a `PlutusData::Bytes(Vec::with_capacity(0))` directly